### PR TITLE
Adding max volumetric flow rate to slow down movements where flow...

### PR DESCRIPTION
**Pull request by mjc-506**
…ratee would exceed the max allowed

# Description

A little additional code to allow CuraEngine to reduce print head velocity to respect any limit set for filament volumetric flow rate (a particular extruder will only be physically capable of melting and depositing so much filament per second, without this patch Cura would command flow rates exceeding this rate at high speeds/layer heights/line thicknesses.)

See https://github.com/smartavionics/Cura/commit/81a6fddfd1d886fe7d69d5ff2d53b787946cbf89?diff=unified for the same changes in 4.20

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Slicing an STL with this PR (and another to Cura to add the settings needed) results in a slightly changed Gcode with reduced print head velocity where the maximum volumetric flow rate would be exceeded.

**Test Configuration**:
* Operating System: Linux

# Checklist:

- [X] My code follows the style guidelines of this project as described in [UltiMaker Meta](https://github.com/Ultimaker/Meta)
- [X] I have read the [Contribution guide](https://github.com/Ultimaker/Cura/blob/main/contributing.md)
- [X] I have commented my code, particularly in hard-to-understand areas
- [NA] I have uploaded any files required to test this change

### CODE REVIEW

**Comment by mashuptwice** [comment]
Tests failed: 

```
/home/runner/work/CuraEngine/CuraEngine/src/LayerPlan.cpp:2129:23: error: assignment of read-only variable ‘extrude_speed’
 2129 |         extrude_speed *= max_mm3_per_sec / mm3_per_sec;
      |         ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

It shouldn't be `const` no?

File: src/LayerPlan.cpp [hunk 1]
```diff
@@ -2020,7 +2028,7 @@ void LayerPlan::writeGCode(GCodeExport& gcode)
                         p0 = p1;
                         gcode.setZ(std::round(z + layer_thickness * length / totalLength));
 
-                        const double extrude_speed = speed * path.speed_back_pressure_factor;
+                        double extrude_speed = speed * path.speed_back_pressure_factor;
```
Comment (casperlamboo) [inline src/LayerPlan.cpp:2031]: Same as before
```suggestion
                        const double extrude_speed = speed * path.speed_back_pressure_factor;
```

**Review by casperlamboo — CHANGES_REQUESTED** [review]
Thank you for your contribution. I think it is a useful feature to add to Cura. I did have some remarks (see comments on the changed lines), please review them and apply.

I do have one other comment. The limiter is applied in two cases. One within the function `writePathWithCoasting` and one in the function `writeGCode`. There is a subtle difference in the implementation between these two functions.

In `writePathWithCoasting` the back pressure compensation is applied before the limiter, while in `writeGCode` the limiter is applied before the back pressure compensation. Is this done deliberately? I think this should be implemented consistently, and I do think the limiter should be applied _after_ the back pressure compensation.

**Comment by mjc-506** [comment]
Good spots, thanks. I'm so confused, I don't remember changing those lines! haha

It seems I've broken my build env again (conan complaining about version requirements...) but I do agree that writePathWithCoasting and writeGCode should be treated the same. Post back pressure compensation makes sense to me too - we're trying to limit the amount of matter passing through the hot end and don't care about toolhead position/velocity/acceleration. Will try get things working my end before moving it just to make sure things still compile afterwards :-)

**Comment by mjc-506** [comment]
but, I think...:

remove lines 1940 - 1946
`            // Extrusion flow rate limiting
            const double mm3_per_sec = path.getExtrusionMM3perMM() * speed;
            if (max_mm3_per_sec > 0 && mm3_per_sec > max_mm3_per_sec)
            {
                speed *= max_mm3_per_sec / mm3_per_sec;
            }
`

and re-insert after lines 1996 and 2031 (substituting extrude_speed for speed, and removing those const's again)

?

**Comment by casperlamboo** [comment]
Could you apply those changed? Un-tested that is. I'll have to test the changes anyways before merging.

**Comment by mjc-506** [comment]
I think I've done it!

**Comment by casperlamboo** [comment]
Hi @mjc-506 sorry for taking this long to get back to you. We have had quite some internal discussions about this feature, and unfortunately we have decided to take a different approach.

We are afraid that, when using this new feature, the print behavior will become unpredictable. Instead we want to introduce a new feature that gives the user more insight whether flow jumps are violated, and can consequently change settings to make sure flow stays within an acceptable range.

Hope you understand our decision.

**Comment by L0laapk3** [comment]
@casperlamboo since this was rejected, what is the proposed workflow to limit the extrusion flow right now? Its unclear to me from your response. This seems like quite a basic feature to me.

I print with mixed line widths in one print and am bottlenecked by my extruder. Please advice me on how to proceed :)

**Comment by L0laapk3** [comment]
@casperlamboo please help me understand your decision.

What is the proposed workflow on hot end limited printers without this feature?

**Comment by looxonline** [comment]
> Hi @mjc-506 sorry for taking this long to get back to you. We have had quite some internal discussions about this feature, and unfortunately we have decided to take a different approach.
> 
> We are afraid that, when using this new feature, the print behavior will become unpredictable. Instead we want to introduce a new feature that gives the user more insight whether flow jumps are violated, and can consequently change settings to make sure flow stays within an acceptable range.
> 
> Hope you understand our decision.

This feels like a decision that is placing a large onus on the user to iterate on their settings until they get it right when in reality most users just want to put a limit in place and know that the slicer will obey it by adjusting the needed settings automatically.

When we consider the approach being taken by other slicers at the moment I feel like it produces great looking results and is exactly what users are looking for. Furthermore, the ability to visualise flow rates on a layer by layer basis still provides the user with the option to manually iterate on the settings even in the other slicers.

**Comment by mjc-506** [comment]
If it's something that won't be included in cura, is this something a plugin could do? I've not even begun to explore that yet, so...

________________________________

**Comment by Arsoth** [comment]
> Hi @mjc-506 sorry for taking this long to get back to you. We have had quite some internal discussions about this feature, and unfortunately we have decided to take a different approach.
> 
> We are afraid that, when using this new feature, the print behavior will become unpredictable. Instead we want to introduce a new feature that gives the user more insight whether flow jumps are violated, and can consequently change settings to make sure flow stays within an acceptable range.
> 
> Hope you understand our decision.

Nope, don't understand your decision at all. Leave the setting as hidden by default if you don't want to "confuse users" (or just disable it if using an UltiMaker brand printer), but not having this setting leaves Cura genuinely mocked in many 3d printing groups I'm in as a slicer that's about as up to date as simplify3d. 

I  *hate* the UI of SS/PS (and OrcaSlicer isn't much better), as they massively suffer from Open Source Software Syndrome, but this is pushing me to wipe Cura from my machine in general as it's falling behind further and further, and refuses to implement features that are considered standard in other slicers. 

Hell, *Bambu Slicer* has max volumetric flow rate. 

Are you really publicly going on record to state your customers are unable to figure out a setting that people buying Bambu Labs printers are able to with no issues (which is the most plug and play printer yet)? Bold move! For April fools day change the UltiMaker logo to a Dunce cap perhaps? 

Sorry if I sound a bit salty, but this decision is baffling and reeks of middle manglement deciding what the customers want! 
(Hell, the work was DONE FOR YOU, essentially all you had to do was merge and test it...)

**Comment by awabom** [comment]
> @casperlamboo please help me understand your teams decision.
> 
> What is the proposed workflow on hot end limited printers without this feature?

I'd like to know this too. Limiting flow rate with the current settings is hard, since Cura can use thicker lines than configured (because of Arachne) and also seems to overspeed sometimes to keep flow rate constant(?). 

The result is having to configure slower speeds than actually needed.

**Comment by PavelStoyan0v** [comment]
The community needs such a feature. We are not idiots so please don't treat us like such. When the feature is not turned on by default it will not ruin anyone's day! Rather it will make many people happy because I too don't like the UIs of other slicers.

**Comment by Arsoth** [comment]
> The community needs such a feature. We are not idiots so please don't treat us like such. When the feature is not turned on by default it will not ruin anyone's day! Rather it will make many people happy because I too don't like the UIs of other slicers.

Hopefully with the engine plugins available now someone can add that (and single top and bottom perimeters!).

I'm a programmer but I don't know this code base or language (Java primarily for myself), otherwise I'd try and do it...

**Comment by N0cturne96** [comment]
Time to change slicer then

> Hi @mjc-506 sorry for taking this long to get back to you. We have had quite some internal discussions about this feature, and unfortunately we have decided to take a different approach.
> 
> We are afraid that, when using this new feature, the print behavior will become unpredictable. Instead we want to introduce a new feature that gives the user more insight whether flow jumps are violated, and can consequently change settings to make sure flow stays within an acceptable range.
> 
> Hope you understand our decision.

Time to change slicer then, shame.

**Comment by Sandman224** [comment]
This is super dumb. Why can't people have the same feature as what's on literally any other slicer? Citing "unpredictable behavior" as a reason for closing this pull request is not only unhelpful but completely unprofessional from a code review standpoint. Clearly from all the duplicate issues regarding this issue this is a feature people want so give some damn constructive feedback and find a way to make it work because people want it!

The proposed solution of notifying the user that they are exceeding a flow rate limit is completely useless because we can already tell that the flow rate limit of an extruder is being exceeded when the filament skips and prints fail, or hell by looking at the print preview. Adding a notification just notifies users of things they already know, while not helping to actually fix the problem. Why should users be required to test and retest numerous settings to get the maximum flow rate they want instead of just capping it, and how that is a more desirable workflow for your slicer is beyond me. 

It seems like the Ultimaker devs are completely out of touch with their community and it's honestly sad. Shame on you for not implementing features that the community wants, and is literally developing for you, **FOR FREE**. @mjc-506 seems more than willing to receive constructive feedback and fix issues with their implementation of flow rate limiting so just provide some actual useful reason why this current implementation is an issue other than it being too naive for your liking.

Again if Ultimaker devs think this is _too confusing_ for users then just have it disabled by default. People want this feature and treating your user base like a bunch of idiots is just completely unacceptable.

In any event I am no longer going to be using Cura after a decade of use. Guess its time to learn a different software.

**Comment by SPYKE357** [comment]
Been waiting for ages for something other slicers have implemented. So long Cura... Time to start transferring my profiles to Prusa. Community wants it. Give it to us. Simple. No stories if's or but's. Been contemplating the move for the last while in the vain hope the people at Ultimaker would comply...

**Comment by Arsoth** [comment]
I'm hoping someone (if I don't get around to it first...) takes mjc-506's code, or at least how it works, and drops it in an engine plugin (and if really lucky, single top/bottom perimeters as well...), so that we can have Cura but good... work's just been too busy for me to code outside of it too lately, so I haven't had a chance to really dig into the engine plugins available from 5.5 onwards.

**Comment by jellespijker** [comment]
@Sandman224 this isn't constructive and affecting the willingness from us to even reconsider it. I will lock the conversation on this PR since I don't accept such a tone.

You can see my answer as given on issue #1990
https://github.com/Ultimaker/CuraEngine/issues/1990#issuecomment-1843787296

As it is now we're not gonna implement such a feature in the engine. If you don't accept such an decision feel free to change to a different slicer @N0cturne96 . 

Maybe someone is willing to Fork Cura and CuraEngine and distribute the binary to whomever is interested. If that developer doesn't want the maintenance burden that comes with maintaining such a fork. I suggest that the effort is instead put into writing a CuraEngine plugin for this.

I'm more then willing to help you guys out there. Feel free to ask for help with such a plugin on this discussion https://github.com/Ultimaker/Cura/discussions/15629 you can tag me there for a quick answer.

**Pull request closed without merging**